crates/sui-framework/packages/sui-framework/sources/accumulator.move

```diff
@@ -17,3 +21,110 @@ fun create(ctx: &TxContext) {
         id: object::sui_accumulator_root_object_id(),
     })
 }
+
+/// Balance object methods
+
+/// Key is used only for computing the field id of accumulator objects.
+/// `T` is the type of the accumulated value, e.g. Balance<SUI>
+public struct Key<phantom T> has copy, drop, store {
+    address: address,
+}
+
+fun root_has_accumulator<K, V: store>(accumulator_root: &AccumulatorRoot, name: Key<K>): bool {
+    dynamic_field::exists_with_type<Key<K>, V>(&accumulator_root.id, name)
+}
+
+use fun root_has_accumulator as AccumulatorRoot.has_accumulator;
+
+fun root_add_accumulator<K, V: store>(
+    accumulator_root: &mut AccumulatorRoot,
+    name: Key<K>,
+    value: V,
+) {
+    dynamic_field::add(&mut accumulator_root.id, name, value);
+}
+
+use fun root_add_accumulator as AccumulatorRoot.add_accumulator;
+
+fun root_borrow_accumulator_mut<K, V: store>(
+    accumulator_root: &mut AccumulatorRoot,
+    name: Key<K>,
+): &mut V {
+    dynamic_field::borrow_mut<Key<K>, V>(&mut accumulator_root.id, name)
+}
+
+use fun root_borrow_accumulator_mut as AccumulatorRoot.borrow_accumulator_mut;
+
+fun root_remove_accumulator<K, V: store>(accumulator_root: &mut AccumulatorRoot, name: Key<K>): V {
+    dynamic_field::remove<Key<K>, V>(&mut accumulator_root.id, name)
+}
+
+use fun root_remove_accumulator as AccumulatorRoot.remove_accumulator;
+
+public(package) fun accumulator_address<T>(address: address): address {
+    let key = Key<T> { address };
+    dynamic_field::hash_type_and_key(sui_accumulator_root_address(), key)
+}
+
+public(package) native fun emit_deposit_event<T>(
+    accumulator: address,
+    recipient: address,
+    amount: u64,
+);
+public(package) native fun emit_withdraw_event<T>(
+    accumulator: address,
+    owner: address,
+    amount: u64,
+);
+
+/// Called by settlement transactions to ensure that the settlement transaction has a unique
+/// digest.
+#[allow(unused_function)]
+fun settlement_prologue(_epoch: u64, _checkpoint_height: u64, _idx: u64, ctx: &TxContext) {
+    assert!(ctx.sender() == @0x0, ENotSystemAddress);
+}
+
+/// A value type for storing any type that is represented in move as a u64.
+/// The additional bits are to prevent overflow, as it would take 2^64 deposits of U64_MAX
+/// to cause an overflow.
+public struct U128 has store {
```

Comment:
Thinking about it more, maybe this should be something like `SumU128`? That is we should encode the operation as well as the size into the name?